File: backend/src/main/java/com/deliveryplatform/service/OrderService.java
package com.deliveryplatform.service;

import com.deliveryplatform.dto.request.CreateOrderRequest;
import com.deliveryplatform.dto.response.OrderResponse;
import com.deliveryplatform.dto.response.PagedResponse;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
import java.time.LocalDate;

public interface OrderService {
    PagedResponse<OrderResponse> getOrders(java.util.UUID driverId, String status, Integer page, Integer size);
    OrderResponse getOrderById(java.util.UUID id, java.util.UUID userId, String role);
    OrderResponse createOrder(CreateOrderRequest request, java.util.UUID clientId);
    OrderResponse updateOrderStatus(java.util.UUID orderId, java.util.UUID userId, String status, Double lat, Double lng, String photoUrl, String scanValue, String comment, Boolean codCollected);
    List<OrderResponse> batchUpdateOrderStatus(List<String> trackingNumbers, String status, Double lat, Double lng, String comment, java.util.UUID userId);
    OrderResponse acceptOrder(java.util.UUID orderId, java.util.UUID userId);
    OrderResponse refuseOrder(java.util.UUID orderId, java.util.UUID userId);
    OrderResponse collectCash(java.util.UUID orderId, java.util.UUID userId);
    OrderResponse reportProblem(java.util.UUID id, String category, String description);
    
    void rateDriver(java.util.UUID orderId, java.util.UUID clientId, Integer rating, String comment);
    List<?> getOrderTracking(java.util.UUID orderId);
    List<OrderResponse> getActiveOrder(java.util.UUID driverId);
    OrderResponse findByTrackingNumberForDriver(String trackingNumber, java.util.UUID driverId);
    OrderResponse findByTrackingNumber(String trackingNumber);
    PagedResponse<OrderResponse> getClientOrders(java.util.UUID clientId, String status, Integer page, Integer size);
    com.deliveryplatform.dto.response.ClientKPIsResponse getClientKPIs(java.util.UUID clientId);
    com.deliveryplatform.dto.response.DriverStatsResponse getDriverStats(java.util.UUID userId);
    void processOrder(java.util.UUID orderId);
    void generateRoute(java.util.UUID orderId);
    
    // Phase 5: Driver Delivery Endpoints
    PagedResponse<OrderResponse> getAvailableOrders(java.util.UUID userId, Integer page, Integer size);
    PagedResponse<OrderResponse> getDriverHistory(java.util.UUID userId, String status, LocalDate startDate, LocalDate endDate, Integer page, Integer size);
    java.util.Map<String, Object> getDriverDashboardStats(java.util.UUID driverId);
    OrderResponse assignOrderToDriver(java.util.UUID orderId, java.util.UUID driverId);
    OrderResponse submitProofOfDelivery(java.util.UUID orderId, java.util.UUID driverId, String proofMethod, String pinCode, MultipartFile photo, String notes) throws Exception;
    OrderResponse confirmPayment(java.util.UUID orderId, java.util.UUID userId);
    void updateSLAStatuses();
    void deleteAllOrders();
}

File: frontend/src/services/api/orderService.ts
import apiClient from '../../api/client';
import { ENDPOINTS } from '../../api/endpoints';
import { Order, PagedResponse } from '../../types';

export interface OrderFilterParams {
  driverId?: string;
  status?: string;
  page?: number;
  size?: number;
}

export interface DriverHistoryParams {
  status?: string;
  startDate?: string;
  endDate?: string;
  page?: number;
  size?: number;
}

export interface StatusUpdateExtra {
  lat?: number;
  lng?: number;
  photoUrl?: string;
  scanValue?: string;
  comment?: string;
  codCollected?: boolean;
  category?: string;
  description?: string;
}

const orderService = {
  getOrders: async (params: OrderFilterParams): Promise<PagedResponse<Order>> => {
    const response = await apiClient.get<PagedResponse<Order>>(ENDPOINTS.ORDERS.BASE, { params });
    return response.data;
  },

  getOrderById: async (id: string): Promise<Order> => {
    const response = await apiClient.get<Order>(ENDPOINTS.ORDERS.BY_ID(id));
    return response.data;
  },

  getOrderDetails: async (id: string): Promise<Order> => {
    return orderService.getOrderById(id);
  },

  updateOrderStatus: async (orderId: string, payload: { status: string } & StatusUpdateExtra): Promise<Order> => {
    const response = await apiClient.put<Order>(ENDPOINTS.ORDERS.UPDATE_STATUS(orderId), payload);
    return response.data;
  },

  acceptOrder: async (orderId: string): Promise<Order> => {
    const response = await apiClient.post<Order>(ENDPOINTS.ORDERS.ASSIGN_DRIVER(orderId));
    return response.data;
  },

  refuseOrder: async (orderId: string): Promise<Order> => {
    const response = await apiClient.post<Order>(ENDPOINTS.ORDERS.REFUSE(orderId));
    return response.data;
  },

  reportProblem: async (orderId: string, category: string, description: string): Promise<Order> => {
    if (orderId.startsWith('fnideq-')) return { id: orderId, status: 'FAILED' } as any;
    const response = await apiClient.post<Order>(ENDPOINTS.ORDERS.REPORT_PROBLEM(orderId), { category, description });
    return response.data;
  },
  
  getDriverKPIs: async (driverId: string): Promise<any> => {
    const response = await apiClient.get(ENDPOINTS.ORDERS.DRIVER.KPIS(driverId));
    return response.data;
  },

  getDriverOrders: async (params: OrderFilterParams): Promise<PagedResponse<Order>> => {
    return orderService.getOrders(params);
  },

  createOrder: async (orderData: any): Promise<Order> => {
    const response = await apiClient.post<Order>(ENDPOINTS.ORDERS.BASE, orderData);
    return response.data;
  },

  batchUpdateStatus: async (data: { trackingNumbers: string[], status: string, lat?: number, lng?: number, comment?: string }): Promise<any> => {
    const response = await apiClient.put(ENDPOINTS.ORDERS.BATCH_STATUS, data);
    return response.data;
  },

  getOrderByTracking: async (trackingNumber: string): Promise<Order> => {
    const response = await apiClient.get<Order>(ENDPOINTS.ORDERS.BY_TRACKING(trackingNumber));
    return response.data;
  },

  getActiveOrders: async (): Promise<Order[]> => {
    const response = await apiClient.get<Order[]>(ENDPOINTS.ORDERS.DRIVER.ACTIVE);
    return Array.isArray(response.data) ? response.data : [];
  },

  estimateFee: async (payload: { distanceKm: number; codAmount?: number; urgent?: boolean; heavy?: boolean }): Promise<any> => {
    const response = await apiClient.post(ENDPOINTS.ORDERS.ESTIMATE_FEE, payload);
    return response.data;
  },

  // Driver-specific methods
  getAvailableOrders: async (params: { page?: number; size?: number; lat?: number; lng?: number; radius?: number } = {}): Promise<PagedResponse<Order>> => {
    const response = await apiClient.get<PagedResponse<Order>>(ENDPOINTS.ORDERS.DRIVER.AVAILABLE, {
      params: { 
        page: params.page || 0, 
        size: params.size || 20,
        lat: params.lat,
        lng: params.lng,
        radius: params.radius
      }
    });
    return response.data;
  },

  getDriverActiveOrders: async (): Promise<Order[]> => {
    const response = await apiClient.get<Order[]>(ENDPOINTS.ORDERS.DRIVER.ACTIVE);
    return response.data;
  },

  getDriverHistory: async (params: DriverHistoryParams): Promise<PagedResponse<Order>> => {
    const response = await apiClient.get<PagedResponse<Order>>(ENDPOINTS.ORDERS.DRIVER.HISTORY, { params });
    return response.data;
  },

  getDriverStats: async (): Promise<any> => {
    const response = await apiClient.get(ENDPOINTS.ORDERS.DRIVER.STATS);
    return response.data;
  },

  assignOrderToDriver: async (orderId: string): Promise<Order> => {
    const response = await apiClient.post<Order>(ENDPOINTS.ORDERS.ASSIGN_DRIVER(orderId));
    return response.data;
  },

  submitProofOfDelivery: async (
    payload: FormData | { orderId: string; proofMethod: string; pinCode?: string; photo?: File; notes?: string }
  ): Promise<Order> => {
    const formData = payload instanceof FormData ? payload : (() => {
      const form = new FormData();
      form.append('orderId', payload.orderId);
      form.append('proofMethod', payload.proofMethod);
      if (payload.pinCode) form.append('pinCode', payload.pinCode);
      if (payload.photo) form.append('photo', payload.photo);
      if (payload.notes) form.append('notes', payload.notes);
      return form;
    })();

    const response = await apiClient.post<Order>(ENDPOINTS.ORDERS.PROOF_OF_DELIVERY, formData);

    return response.data;
  },

  rateDriver: async (orderId: string, rating: number, comment?: string): Promise<{ message: string }> => {
    const response = await apiClient.post(ENDPOINTS.ORDERS.RATE(orderId), { rating, comment });
    return response.data;
  },

  getPublicTracking: async (trackingNumber: string): Promise<Order> => {
    const response = await apiClient.get<Order>(`/public/tracking/${trackingNumber}`);
    return response.data;
  }
};

export default orderService;
